## Predicting sentiment by topic : A intelligent restaurant review

### Process - Intelligent system
#### All reviews for restaurant -> Break all reviews into sentences -> Sentiment classifier -> Output (positive / negative)
#### All reviews -> Break all reviews into sentences -> select sentences with what u looking for (Sushi) -> Sentence sentiment classifier -> Average predictions

### Examples of other classifier
### Classifier :
#### Takes input x -> Classifier MODEL -> output : y  - predicted class 
### Applications
* Consider a webpage and classify its content -> Education, finance, technology, etc; (Input - webpage, output -> category of information).
* Spam filtering : email -> spam filter -> classifies (spam / not-spam);
* Computer Vision : input image  -> classifier -> predict (type of image);
* Personalized medical diagnosis : human parameter -> human classifier model -> medical diagnosis output
* Reading your mind: (FMRI): input (human brain) -> reading -> output(Hammer / house, etc);

### Linear Classifier
#### In sentiment classifier -> a "simple threshold classifier" is used;
                     | List of positive words | List of negative words |
                     ----------------------------------------------------
                                             |
                                            \ /
                     ----------------------------------------------------                      
                     |               Simple threshold classifier        |
                     | * Count positive & begative words in sentence.   |
                     | * If # positive words > # negative words         |
                     |      output (y hat) = positive;                  |
    Sentence         |   else                                           |
    from       --->  |       output (y hat) = negative;                 |
    review           |                                                  |
    (input x)        |                                                  |
                     |                                                  |
                     |                                                  |        
                     |--------------------------------------------------|
                     
#### Problems with threshold classifier
* How to list positive / negative words ?
* Words have different degrees of sentiment: How do we weigh different words?
* Single words are not good enough: E.g : good -> positive ; not good -> negative;

#### The first two issues can be resolved by learning a classifier;
#### The third issue is addressed by more elaborate features.


### Linear Classifier
* It takes all the words and adds weights to them.
* Example

                Words                Weights
                ----------------------------
                good                   1.0
                great                  1.5
                awesome                2.7
                bad                   -1.0  
                terrible              -2.1
                aweful                -3.3
                restaurant,we,the,etc  0.0

* Scoring a sentence.
* Example:
* Input x -> Sushi was "great" , the food was "awesome", but the service was "terrible".
* Score (x) = 1.5 + 2.7 - 2.1 = 2.1 
* Score(x) > 0 = positive review;
* Output is the weight of the input

                    |            WORDS       |          WEIGHTS         |
                     ----------------------------------------------------
                                             |
                                            \ /
                     ----------------------------------------------------                      
                     |               Simple LINEAR classifier           |
                     | * Score(x)= weighted count of words in sentence. |
                     | * If Score(x) > 0 :                              |
                     |      output (y hat) = positive;                  |
      Sentence       |   else                                           |
      from     --->  |       output (y hat) = negative;                 |
      review         |                                                  |
      (input x)      |                                                  |
                     |                                                  |
                     |                                                  |        
                     |--------------------------------------------------|
                     

### Decision Boundary:
* Classifier attempt to make decisions.
* Decision boundaries : Score(x) = weight of word1 * # occurence of the word1 + weight of word2 * # occurence of the word2 + ...
* A plot based of the various words and their ocuurence can be visualized.
* A prediction can be made on the plot an dline segregates the positive from the negative reviews.
* The classifier / decision boundary / line of segregation is 0 and therefore is termed the linear classifier, linear decision boundary.

#### For linear classifiers:
* When 2 weights are non-zero -> line
* When 3 weights are non-zero -> plane (positive prediction are above the plane ; negative predictions are below the plane);
* When many weights are non-zer0 -> hyperplanes

##### For more generale classifiers (not linear) - more complicated shapes - segregate the data.


### Evaluating Classification models
#### Training and evaluating a classifier
* In regression model -> the error - rss - residual sum of squares
* In classification model -> error are different;

### In classification - Training a classifier

* Learning the weight

#### Data (x,y) [(sentence1, +), (sentence2, -),..] -> split 
                                             -> training set -> learn classifier(words|weights)
                                                                      |
                                                                     \ /
                                             ->  test set ----------> evaluate ?   
                                             

### Evaluation - Classification error:
* Testexample -> Learned classifier -> output predicted ; test error - evalutes the predicted output w.r.t true label -> if its a match then its correct else it is false;

#### Examples
 * Review (Sushi was great, +) -> sentence : Sushi was great, true label : +;
 * the sentence "Sushi was great" is feed to the learned classifier; while he true label "+" is hidden;
 * The classifier predicts the output. In this case -> +
 * Since the predicted output = + , the true label = + ; its a match - therefore Correct;
 * -------------------
 * Review (Food was OK, -) -> Sentence : Food was OK, true label : -;
 * the sentence "Food was ok" is feed to the learned classifier; while the true label "-" is hidden;
 * the classifier predicts the output. In this case -> +
 * Since the predicted output = + , the true label = - ; its a mis-match - therefore Mistake;
 
##### The process is repeated for all examples.
 
### Classification Error and  Accuracy:
 * Error measures the fraction of mistake  = # mistakes / total # of sentences;
 * Best possible value of error = 0.0;
 * error = 1 - accuracy;
 
 * Accuracy  measure the fraction of correct predictions = # correct / total # of sentences;
 * Best possible value of accuracy = 1.0;
 * accuracy = 1 - error;
 
#### For Good Accuracy -> prevent random guessing
* for binary classification - on average - accuracy = 0.5 ,  rest random guessing;
* for k class, accuracy = 1/k ; 0.333 for 3 classes, 0.25  for 4 classes, etc;

#### Is a classifier with 90% accuracy good? 
* Depends
* In case of spam - not good - case of major class prediction - It has amazing performance when there is a "class imbalance" (One class is more common than another, beats random (if you know the majority class));

#### Need to know what accuracy does the application need.


### Confusion Matrix - Types of error

* The relationship between the true label and the predicted label - is termed confudion matrix.

                            Predicted Label
                          |---------|---------|
                          |         |         |
                          |     +   |    -    |
                          |         |         |
                  |-------|---------|---------|   
            True  |       | True    |   False |
            Label |   +   | Positive| Negative|
                  |       |         |         |
                  |-------|---------|---------|
                  |       | False   |   True  |
                  |    -  | Positive| Negative| 
                  |       |         |         |
                  |----------------------------
                  
### Cost of different types of mistakes can be different (and high) in some applications
                          |---------|---------|
                          | Spam    | Medical |
                          |Filtering|Diagnosis|
                          |         |         |
                 |--------|---------|---------|   
                 | False  |         | Disease |
                 |Positive| Annoying|   not   |
                 |        |         | treated |<-----|
                 |--------|---------|---------|    high cost
                 |False   | Email   | Wasteful|      |
                 |Negative|  Lost   |treatment|<-----|
                 |        |high cost|         |
                 |--------|-------------------|

### Confudion matrix - binary classification
* 100 test examples - 60 are positive (true); 40 are negative (true);


                               Predicted Label
                             |---------|---------|
                             |         |         |       Accuracy = 85 / 100 = 0.85
                             |     +   |    -    |       False Negative = 10 / 100 = 0.1
                             |         |         |       False Positive = 5 / 100 = 0.05
                  |----------|---------|---------|       FN > FP;
            True  |          |correct  |         |
            Label |   + 60   |   50    |    10   |
                  |          |         |         |
                  |----------|---------|---------|
                  |          |         |correct  |
                  |    - 40  |    5    |    35   | 
                  |          |         |         |
                  |-------------------------------

### Confusion matrix - multiclass classification
* 100 test examples - 70 healthy, 20 cold, 10 flu;

                                    Predicted Label
                          |---------|---------|--------
                          |         |         |       |
                          | Healthy |    Cold | Flu   |
                          |         |         |       |
                  |-------|---------|---------|--------   
            True  |       | correct |         |       |    Accuracy = 80 /100 = 0.8
            Label |Healthy|    60   |     8   |   2   | 
                  |  70   |         |         |       |
                  |-------|---------|---------|--------
                  |       |         | correct |       |
                  |  Cold |    4    |     12  |   4   |
                  |   20  |         |         |       |
                  -------------------------------------
                  |       |         |         |correct|
                  |  Flu  |    0    |     2   |   8   |
                  |  10   |         |         |       |
                  |------------------------------------
                  
                  


### Learning Curves

#### How much data does a model need to learn?
* The more the better - but data quality is an important factor.
* Theoretical techniques can bound the amount of data needed - typically loose for practical application, but provides guidance.
* In practice - more complex models require more data; empirical analysis can provide guidance.

#### Learning curves - relates the amount of training data and the test error.
* If the amount of training data is high, then the test error will be low ;
* If the amount of training data is low , then the test error will be high;
* Each model has a gap - bias, beyond which the although the number of training data increases the test error will not decrease.

#### More complex models tend to have less bias.
* Sentiment classifier using single words can do Ok;
* Less bias -> potentially more accurate, needs more data to learn.

##### Models with less bias tends to need more data to learn well, but to do better with sufficient data. 

### Class Probabilities
* Since all the reviews, classification can't just fit into 2 categories; 
* Many classifiers provide a confidencec level, extremely useful in practice.
* P(y|x) -> y: output label ; x : input sentence ;

### Classification Model
                                    
       
       Training Data           Feature        X (word count)    ML   y hat(predicted sentiment)
       (Test st review, --->   Extraction -------------------> model -------------------------->
       sentiment)                                               / \                         |
          |                                                      |                          | 
          | y (sentiment label)                                 w hat (weight for each word)|  
          |                                                     / \                         |
          |                                                      |                          |
          |                                                      |                          |
          |                                                   ML Algorithm                  |
          |                                                     / \                         |
          |                                                      |  classification accuracy |
          |                                                      |                          |
          |                                        y          Quality        y hat          |
          |--------------------------------------------------> Metric <---------------------|
                                                               

### Analyzing product sentiment

In [1]:
import graphlab

#### Read some product review data

In [2]:
products = graphlab.SFrame('amazon_baby.gl/')

This non-commercial license of GraphLab Create for academic use is assigned to amitha353@gmail.com and will expire on May 07, 2019.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: C:\Users\Amitha\AppData\Local\Temp\graphlab_server_1526133099.log.0


#### Explore the data 

In [3]:
products.head()

Columns:
	name	str
	review	str
	rating	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[10 rows x 3 columns]

### Create a word count vector for each review
* take the review -> separate words (tokenizing) -> build count vector; all this done by tools.

In [4]:
products['word_count'] = graphlab.text_analytics.count_words(products['review'])

In [5]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'and': 5L, 'stink': 1L, '... |
| {'and': 3L, 'love': 1L, 'i... |
| {'and': 2L, 'quilt': 1L, '... |
| {'ingenious': 1L, 'and': 3... |
| {'and': 2L, 'parents!!': 1... |
| {'and': 2L, 'cute': 1L, 'h... |
| {'shop': 1L, 'be': 1L, 'is... |
| {'feeding,': 1L, 'and': 2L... |
| {'and': 1L, 'help': 1L, 'g... |
| {'journal.': 1L, 'all': 1L... |
+-------------------------------+
[10 rows x 4 columns]

In [6]:
graphlab.canvas.set_target('ipynb')

In [7]:
products['name'].show() # histogram of the products

### Explore the most frequently reviewed product - Vulli Sophie

In [8]:
giraffe_reviews = products[products['name'] == 'Vulli Sophie the Giraffe Teether']

In [9]:
len(giraffe_reviews)

785

In [10]:
giraffe_reviews['rating'].show(view='Categorical')

### Build a sentiment classifier

In [11]:
products['rating'].show(view='Categorical')

### Defining positive (4,5) and negative (1,2) reviews; 3 is removed -> data engineering

In [12]:
products = products[products['rating'] != 3]

In [13]:
products['sentiment'] = products['rating'] >=4

In [14]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 3L, 'love': 1L, 'i... |     1     |
| {'and': 2L, 'quilt': 1L, '... |     1     |
| {'ingenious': 1L, 'and': 3... |     1     |
| {'and': 2L, 'parents!!': 1... |     1     |
| {'and': 2L, 'cute': 1L, 'h... |     1     |
| {'shop': 1L, 'be': 1L, 'is... |     1     |
| {'feeding,': 1L, 'and': 2L... |     1     |
| {'and': 1L, 'help': 1L, 'g... |     1     |
| {'journal.': 1L, 'all': 1L... |     1     |
| {'all': 1L, 'forget': 1L, ... |     1     |
+-------------------------------+-----------+
[10 rows x 5 columns]

### Training the sentiment classifier

In [15]:
train_data, test_data = products.random_split(.8, seed=0)

In [16]:
sentiment_model = graphlab.logistic_classifier.create(train_data, 
                                                      target='sentiment', 
                                                      features = ['word_count'], 
                                                      validation_set=test_data)

WARNING: The number of feature dimensions in this problem is very large in comparison with the number of examples. Unless an appropriate regularization value is set, this model may not provide accurate predictions for a validation/test set.

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 219217

Number of coefficients    : 219218

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 5        | 0.000002  | 4.177602     | 0.841481          | 0.839989            |

| 2         | 9        | 3.000000  | 6.069229     | 0.947425          | 0.894877            |

| 3         | 10       | 3.000000  | 6.717446     | 0.923768          | 0.866232            |

| 4         | 11       | 3.000000  | 7.314231     | 0.971779          | 0.912743            |

| 5         | 12       | 3.000000  | 7.839167     | 0.975511          | 0.908900            |

| 6         | 13       | 3.000000  | 8.340501     | 0.899991          | 0.825967            |

| 10        | 18       | 1.000000  | 10.551148    | 0.988715          | 0.916256            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

### Evaluate the sentiment model

In [17]:
sentiment_model.evaluate(test_data, metric = 'roc_curve') # roc_curve - explore confusion matric, false positive, false negatives;

{'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+----------------+----------------+-------+------+
 | threshold |      fpr       |      tpr       |   p   |  n   |
 +-----------+----------------+----------------+-------+------+
 |    0.0    |      1.0       |      1.0       | 27976 | 5328 |
 |   1e-05   | 0.909346846847 | 0.998856162425 | 27976 | 5328 |
 |   2e-05   | 0.896021021021 | 0.998748927652 | 27976 | 5328 |
 |   3e-05   | 0.886448948949 | 0.998462968259 | 27976 | 5328 |
 |   4e-05   | 0.879692192192 | 0.998284243637 | 27976 | 5328 |
 |   5e-05   | 0.875187687688 | 0.998212753789 | 27976 | 5328 |
 |   6e-05   | 0.872184684685 | 0.998177008865 | 27976 | 5328 |
 |   7e-05   | 0.868618618619 | 0.998034029168 | 27976 | 5328 |
 |   8e-05   | 0.864677177177 | 0.997998284244 | 27976 | 5328 |
 |   9e-05   | 0.860735735736 | 0.997962539319 | 27976 | 5328 |
 +-----------+----------------+----------------+-------+------

In [18]:
sentiment_model.show(view='Evaluation')

### Apply the learned model to understand the product giraffe

In [19]:
giraffe_reviews['predicted_sentiment'] = sentiment_model.predict(giraffe_reviews, output_type='probability')

In [20]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | He likes chewing on all th... |  5.0   |
| Vulli Sophie the Giraffe T... | My son loves this toy and ... |  5.0   |
| Vulli Sophie the Giraffe T... | There really should be a l... |  1.0   |
| Vulli Sophie the Giraffe T... | All the moms in my moms' g... |  5.0   |
| Vulli Sophie the Giraffe T... | I was a little skeptical o... |  5.0   |
| Vulli Sophie the Giraffe T... | I have been reading about ... |  5.0   |
| Vulli Sophie the Giraffe T... | My neice loves her sophie ... |  5.0   |
| Vulli Sophie the Giraffe T... | What a friendly face!  And... |  5.0   |
| Vulli Sophie the Giraffe T... | We got this just for my so... |  5.0   |
| Vulli Sophie the Giraffe T... | My baby seems to like this... |  3.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------------------+
|           word_count          | predicted_sentiment |
+-------------------------------+---------------------+
| {'and': 1L, 'all': 1L, 'be... |    0.999513023521   |
| {'and': 1L, 'right': 1L, '... |    0.999320678306   |
| {'and': 2L, 'all': 1L, 'la... |    0.013558811687   |
| {'and': 2L, 'one!': 1L, 'a... |    0.995769474148   |
| {'and': 3L, 'all': 1L, 'ol... |    0.662374415673   |
| {'and': 6L, 'seven': 1L, '... |    0.999997148186   |
| {'and': 4L, 'drooling,': 1... |    0.989190989536   |
| {'and': 3L, 'chew': 1L, "d... |    0.999563518413   |
| {'chew': 2L, 'because': 1L... |    0.970160542725   |
| {'and': 2L, 'already': 1L,... |    0.195367644588   |
+-------------------------------+---------------------+
[10 rows x 5 columns]

### Sort the reviews based on the predicted sentiment and explore

In [21]:
giraffe_reviews = giraffe_reviews.sort('predicted_sentiment', ascending=False)

In [22]:
giraffe_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Vulli Sophie the Giraffe T... | Sophie, oh Sophie, your ti... |  5.0   |
| Vulli Sophie the Giraffe T... | I'm not sure why Sophie is... |  4.0   |
| Vulli Sophie the Giraffe T... | I'll be honest...I bought ... |  4.0   |
| Vulli Sophie the Giraffe T... | We got this little giraffe... |  5.0   |
| Vulli Sophie the Giraffe T... | As a mother of 16month old... |  5.0   |
| Vulli Sophie the Giraffe T... | Sophie the Giraffe is the ... |  5.0   |
| Vulli Sophie the Giraffe T... | Sophie la giraffe is absol... |  5.0   |
| Vulli Sophie the Giraffe T... | My 5-mos old son took to t... |  5.0   |
| Vulli Sophie the Giraffe T... | My nephews and my four kid... |  5.0   |
| Vulli Sophie the Giraffe T... | Never thought I'd see my s... |  5.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------------------+
|           word_count          | predicted_sentiment |
+-------------------------------+---------------------+
| {'giggles': 1L, 'all': 1L,... |         1.0         |
| {'peace': 1L, 'month': 1L,... |    0.999999999703   |
| {'all': 2L, 'pops': 1L, 'e... |    0.999999999392   |
| {'all': 2L, "don't": 1L, '... |    0.99999999919    |
| {'cute': 1L, 'all': 1L, 'r... |    0.999999998657   |
| {'just': 2L, 'both': 1L, '... |    0.999999997108   |
| {'and': 5L, 'the': 1L, 'al... |    0.999999995589   |
| {'just': 1L, 'shape': 2L, ... |    0.999999995573   |
| {'and': 4L, 'chew': 1L, 'a... |    0.999999989527   |
| {'giggles': 1L, 'all': 1L,... |    0.999999985069   |
+-------------------------------+---------------------+
[10 rows x 5 columns]

In [23]:
giraffe_reviews[0]['review']

"Sophie, oh Sophie, your time has come. My granddaughter, Violet is 5 months old and starting to teeth. What joy little Sophie brings to Violet. Sophie is made of a very pliable rubber that is sturdy but not tough. It is quite easy for Violet to twist Sophie into unheard of positions to get Sophie into her mouth. The little nose and hooves fit perfectly into small mouths, and the drooling has purpose. The paint on Sophie is food quality.Sophie was born in 1961 in France. The maker had wondered why there was nothing available for babies and made Sophie from the finest rubber, phthalate-free on St Sophie's Day, thus the name was born. Since that time millions of Sophie's populate the world. She is soft and for babies little hands easy to grasp. Violet especially loves the bumpy head and horns of Sophie. Sophie has a long neck that easy to grasp and twist. She has lovely, sizable spots that attract Violet's attention. Sophie has happy little squeaks that bring squeals of delight from Viol

In [24]:
giraffe_reviews[1]['review']

"I'm not sure why Sophie is such a hit with the little ones, but my 7 month old baby girl is one of her adoring fans.  The rubber is softer and more pleasant to handle, and my daughter has enjoyed chewing on her legs and the nubs on her head even before she started teething.  She also loves the squeak that Sophie makes when you squeeze her.  Not sure what it is but if Sophie is amongst a pile of her other toys, my daughter will more often than not reach for Sophie.  And I have the peace of mind of knowing that only edible and safe paints and materials have been used to make Sophie, as opposed to Bright Starts and other baby toys made in China.  Now that the research is out on phthalates and other toxic substances in baby toys, I think it's more important than ever to find good quality toys that are also safe for our babies to handle and put in their mouths.  Sophie is a must-have for every new mom in my opinion.  Even if your kid is one of the few that can take or leave her, it's worth

### Show most negative review

In [25]:
giraffe_reviews[-1]['review']

"My son (now 2.5) LOVED his Sophie, and I bought one for every baby shower I've gone to. Now, my daughter (6 months) just today nearly choked on it and I will never give it to her again. Had I not been within hearing range it could have been fatal. The strange sound she was making caught my attention and when I went to her and found the front curved leg shoved well down her throat and her face a purply/blue I panicked. I pulled it out and she vomited all over the carpet before screaming her head off. I can't believe how my opinion of this toy has changed from a must-have to a must-not-use. Please don't disregard any of the choking hazard comments, they are not over exaggerated!"

In [26]:
giraffe_reviews[-2]['review']

"This children's toy is nostalgic and very cute. However, there is a distinct rubber smell and a very odd taste, yes I tried it, that my baby did not enjoy. Also, if it is soiled it is extremely difficult to clean as the rubber is a kind of porus material and does not clean well. The final thing is the squeaking device inside which stopped working after the first couple of days. I returned this item feeling I had overpaid for a toy that was defective and did not meet my expectations. Please do not be swayed by the cute packaging and hype surounding it as I was. One more thing, I was given a full refund from Amazon without any problem."

## Assignment

### 1. Use .apply() to build a new feature with the counts for each of the selected_words

In [27]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

In [28]:
def word_count(word, word_dict):
    if word in word_dict:
        return word_dict[word];
    else:
        return 0;

In [29]:
for word in selected_words:
    products[word] = products['word_count'].apply(lambda word_dict : word_count(word, word_dict))

In [35]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'and': 3L, 'love': 1L, 'i... |     1     |    0    |   0   |     0     |    0    |
| {'and': 2L, 'quilt': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'ingenious': 1L, 'and': 3... |     1     |    0    |   0   |     0     |    0    |
| {'and': 2L, 'parents!!': 1... |     1     |    0    |   1   |     0     |    0    |
| {'and': 2L, 'cute': 1L, 'h... |     1     |    0    |   1   |     0     |    0    |
| {'shop': 1L, 'be': 1L, 'is... |     1     |    0    |   0   |     0     |    0    |
| {'feeding,': 1L, 'and': 2L... |     1     |    0    |   0   |     0     |    0    |
| {'and': 1L, 'help': 1L, 'g... |     1     |    0    |   0   |     0     |    0    |
| {'journal.': 1L, 'all': 1L... |     1     |    0    |   0   |     0     |    0    |
| {'all': 1L, 'forget': 1L, ... |     1     |    0    |   0   |     0     |    0    |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+
| love | horrible | bad | terrible | awful | wow | hate |
+------+----------+-----+----------+-------+-----+------+
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |
+------+----------+-----+----------+-------+-----+------+
[10 rows x 16 columns]

In [45]:
sumlist = []
for word in selected_words:
    temp = products[word].sum()
    print "%s : %i" %(word, products[word].sum())
    sumlist.append(temp)

awesome : 2002
great : 42420
fantastic : 873
amazing : 1305
love : 40277
horrible : 659
bad : 3197
terrible : 673
awful : 345
wow : 131
hate : 1057


### Q1 . Out of the 11 words in selected_words, which one is most used in the reviews in the dataset?

In [46]:
max(sumlist)  # great

42420L

### Q2 . Out of the 11 words in selected_words, which one is least used in the reviews in the dataset? 

In [47]:
min(sumlist)  # wow

131L

### 2.  Create a new sentiment analysis model using only the selected_words as features:

In [32]:
train_data_set,test_data_set = products.random_split(.8, seed=0)

In [39]:
selected_words_model = graphlab.logistic_classifier.create(train_data_set,
                                                          target = 'sentiment',
                                                          features = selected_words,
                                                          validation_set=test_data_set)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients    : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.374671     | 0.844299          | 0.842842            |

| 2         | 3        | 0.691902     | 0.844186          | 0.842842            |

| 3         | 4        | 0.951610     | 0.844276          | 0.843142            |

| 4         | 5        | 1.163729     | 0.844269          | 0.843142            |

| 5         | 6        | 1.371460     | 0.844269          | 0.843142            |

| 6         | 7        | 1.569301     | 0.844269          | 0.843142            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [34]:
# find the coefficient - most positive and negative of the selected word
selected_words_model = selected_words_model['coefficients'].sort('value')

In [36]:
selected_words_model.print_rows(num_rows=15)

+-------------+-------+-------+------------------+------------------+
|     name    | index | class |      value       |      stderr      |
+-------------+-------+-------+------------------+------------------+
|   terrible  |  None |   1   |  -2.09049998487  | 0.0967241912229  |
|   horrible  |  None |   1   |  -1.99651800559  | 0.0973584169028  |
|    awful    |  None |   1   |  -1.76469955631  |  0.134679803365  |
|     hate    |  None |   1   |  -1.40916406276  | 0.0771983993506  |
|     bad     |  None |   1   | -0.985827369929  | 0.0433603009142  |
|     wow     |  None |   1   | -0.0541450123333 |  0.275616449416  |
|    great    |  None |   1   |  0.883937894898  | 0.0217379527921  |
|  fantastic  |  None |   1   |  0.891303090304  |  0.154532343591  |
|   amazing   |  None |   1   |  0.892802422508  |  0.127989503231  |
|   awesome   |  None |   1   |  1.05800888878   |  0.110865296265  |
| (intercept) |  None |   1   |  1.36728315229   | 0.00861805467824 |
|     love    |  Non

### Q3 : Out of the 11 words in selected_words, which one got the most positive weight in the selected_words_model?

### love

### Q4 : Out of the 11 words in selected_words, which one got the most negative weight in the selected_words_model?

### terrible

### 3. Comparing the accuracy of different sentiment analysis model

### Q6 : Which of the following ranges contains the accuracy of the sentiment_model in the IPython Notebook from lecture on the test_data?

In [37]:
sentiment_model.evaluate(test_data_set)

{'accuracy': 0.916256305548883,
 'auc': 0.9446492867438502,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      1       |        0        |  1461 |
 |      0       |        1        |  1328 |
 |      0       |        0        |  4000 |
 |      1       |        1        | 26515 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9500349343413533,
 'log_loss': 0.2610669843242193,
 'precision': 0.9523039902309378,
 'recall': 0.9477766657134686,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+----------------+----------------+-------+------+
 | threshold |      fpr       |      tpr       |   p   |  n   |
 +-----------+----------------+----------------+-------+------+
 |    0.0    |      1.0       |  

### 0.901 to 0.931

### Q5 :  Which of the following ranges contains the accuracy of the selected_words_model on the test_data?

In [40]:
selected_words_model.evaluate(test_data_set)

{'accuracy': 0.8431419649291376,
 'auc': 0.6648096413721418,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      0       |        0        |  234  |
 |      1       |        0        |  130  |
 |      0       |        1        |  5094 |
 |      1       |        1        | 27846 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.914242563530107,
 'log_loss': 0.4054747110366022,
 'precision': 0.8453551912568306,
 'recall': 0.9953531598513011,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+-----+-----+-------+------+
 | threshold | fpr | tpr |   p   |  n   |
 +-----------+-----+-----+-------+------+
 |    0.0    | 1.0 | 1.0 | 27976 | 5328 |
 |   1e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   2e-05   | 

### 0.841 to 0.871

In [56]:
selected_words_model.evaluate(test_data_set, metric='roc_curve')

{'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+-----+-----+-------+------+
 | threshold | fpr | tpr |   p   |  n   |
 +-----------+-----+-----+-------+------+
 |    0.0    | 1.0 | 1.0 | 27976 | 5328 |
 |   1e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   2e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   3e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   4e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   5e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   6e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   7e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   8e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   9e-05   | 1.0 | 1.0 | 27976 | 5328 |
 +-----------+-----+-----+-------+------+
 [100001 rows x 5 columns]
 Note: Only the head of the SFrame is printed.
 You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.}

In [57]:
selected_words_model.show(view='Evaluation')

### Q7 : Which of the following ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the test_data?

* Majority class is the highest probability of an event in a process.
* Here reviews with rating 1, 2 are negative (sentiment = 0); 
* Reviews with rating 4, 5 are positive (sentiment 1);
* Reviews with rating 3 are discarded.
* There are 2 classes - sentiment - 1 or 0; Occurence -> Which ever occurs more frequently, its the majority class.

In [64]:
total = float(test_data_set.num_rows())

In [65]:
class_1 = float(test_data_set[test_data_set['sentiment'] == 1].num_rows())

In [69]:
class_2 = float(test_data_set[test_data_set['sentiment'] == 0].num_rows())

In [67]:
if class_1 > class_2:
    maj_class = class_1;
else:
    maj_class = class_2;

In [70]:
print maj_class / total

0.840019216911


### 0.811 to 0.843

### Q8 : How do you compare the different learned models with the baseline approach where we are just predicting the majority class? 

### The model learned using all words performed much better than the other two. The other two approaches performed about the same.
* selected_words_model.evaluate(test_data_set) # 0.843
* selected_words_model.evaluate(test_data_set, metric='roc_curve') # 0.843
* sentiment_model.evaluate(test_data_set) # 0.916

### 4. Interpreting the difference in performance between the models: 

### Q9 : Which of the following ranges contains the ‘predicted_sentiment’ for the most positive review for ‘Baby Trend Diaper Champ’, according to the sentiment_model from the IPython Notebook from lecture?

In [43]:
diaper_champ_reviews = products[products['name'] == 'Baby Trend Diaper Champ']

In [44]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'just': 2L, 'less': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'just': 1L, 'less': 1L, '... |     0     |    0    |   0   |     0     |    0    |
| {'control': 1L, 'am': 1L, ... |     1     |    0    |   0   |     0     |    0    |
| {'and': 3L, 'over.': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'just': 1L, 'when': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'lysol': 1L, 'all': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'all': 1L, 'bags.': 1L, '... |     0     |    0    |   0   |     0     |    0    |
| {'lysol': 1L, 'all': 2L, '... |     1     |    0    |   0   |     0     |    0    |
| {'and': 2L, 'all': 1L, 'ju... |     1     |    0    |   0   |     0     |    0    |
| {'hand.': 1L, '(required':... |     1     |    0    |   0   |     0     |    0    |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+
| love | horrible | bad | terrible | awful | wow | hate |
+------+----------+-----+----------+-------+-----+------+
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    1     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  1  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |
+------+----------+-----+----------+-------+-----+------+
[10 rows x 16 columns]

In [48]:
diaper_champ_reviews['rating'].show(view='Categorical')

In [49]:
diaper_champ_reviews['predicted_sentiment'] = sentiment_model.predict(diaper_champ_reviews, output_type='probability')

In [50]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

In [51]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Baby Luke can turn a clean... |  5.0   |
| Baby Trend Diaper Champ | I LOOOVE this diaper pail!... |  5.0   |
| Baby Trend Diaper Champ | We researched all of the d... |  4.0   |
| Baby Trend Diaper Champ | My baby is now 8 months an... |  5.0   |
| Baby Trend Diaper Champ | This is absolutely, by far... |  5.0   |
| Baby Trend Diaper Champ | Diaper Champ or Diaper Gen... |  5.0   |
| Baby Trend Diaper Champ | Wow!  This is fabulous.  I... |  5.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I am one of those super-cr... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'all': 1L, 'less': 1L, "f... |     1     |    0    |   0   |     0     |    0    |
| {'just': 1L, 'over': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'all': 2L, 'just': 4L, "d... |     1     |    0    |   0   |     0     |    0    |
| {"don't": 1L, 'when': 1L, ... |     1     |    0    |   2   |     0     |    0    |
| {'just': 3L, 'money': 1L, ... |     1     |    0    |   0   |     0     |    0    |
| {'all': 1L, 'bags.': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'and': 4L, '"genie".': 1L... |     1     |    0    |   0   |     0     |    0    |
| {'lysol': 1L, 'all': 2L, '... |     1     |    0    |   0   |     0     |    0    |
| {'just': 1L, 'when': 1L, '... |     1     |    0    |   0   |     0     |    0    |
| {'taller': 1L, 'bags.': 1L... |     1     |    0    |   0   |     0     |    0    |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999937267   |
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999917406   |
|  0   |    0     |  1  |    0     |   0   |  0  |  0   |    0.999999899509   |
|  0   |    0     |  1  |    0     |   0   |  0  |  0   |    0.999999836182   |
|  2   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999824745   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999759315   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999692111   |
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999642488   |
|  0   |    1     |  0  |    0     |   0   |  0  |  0   |    0.999999604504   |
|  1   |    0     |  0  |    0     |   0   |  0  |  0   |    0.999999486804   |
+------+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

### Range : 0.9 - 1.0

### Q 10: Consider the most positive review for ‘Baby Trend Diaper Champ’ according to the sentiment_model from the IPython Notebook from lecture. Which of the following ranges contains the predicted_sentiment for this review, if we use the selected_words_model to analyze it?

In [52]:
diaper_champ_reviews['predicted_sentiment'] = selected_words_model.predict(diaper_champ_reviews, output_type='probability')

In [53]:
diaper_champ_reviews = diaper_champ_reviews.sort('review', ascending=True)

In [54]:
selected_words_model.predict(diaper_champ_reviews[0:1], output_type='probability')

dtype: float
Rows: 1
[0.796940851290673]

### Range : 0.7 - 0.8

### Question 11 : Why is the value of the predicted_sentiment for the most positive review found using the sentiment_model much more positive than the value predicted using the selected_words_model?

In [55]:
# Most positive review = Baby Luke can turn a ...
diaper_champ_reviews[(diaper_champ_reviews['review'].contains('Baby Luke can turn a'))]

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	predicted_sentiment	float

Rows: Unknown

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Baby Luke can turn a clean... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+---------+-------+-----------+---------+
|           word_count          | sentiment | awesome | great | fantastic | amazing |
+-------------------------------+-----------+---------+-------+-----------+---------+
| {'all': 1L, 'less': 1L, "f... |     1     |    0    |   0   |     0     |    0    |
+-------------------------------+-----------+---------+-------+-----------+---------+
+------+----------+-----+----------+-------+-----+------+---------------------+
| love | horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+------+----------+-----+----------+-------+-----+------+---------------------+
|  0   |    0     |  0  |    0     |   0   |  0  |  0   |    0.796940851291   |
+------+----------+-----+----------+-------+-----+------+---------------------+
[? rows x 17 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

### None of the selected_words appeared in the text of this review. 